logbook-servlet/src/main/java/org/zalando/logbook/servlet/RemoteRequest.java

```diff
@@ -1,36 +1,192 @@
+            if (isFormRequest(request)) {
+                switch (formRequestMode) {
+                    case PARAMETER:
+                        return new Buffering(reconstructFormBody(request));
+                    case OFF:
+                        return new Passing();
+                    default:
+                        break;
+                }
+            }
+
+            return new Buffering(toByteArray(request.getInputStream()));
+        }
+
+        private boolean isFormRequest(final ServletRequest request) {
+            return Optional.ofNullable(request.getContentType())
+                    .flatMap(MimeTypes::parse)
+                    .filter(this::isFormRequest)
+                    .isPresent();
+        }
+
+        private boolean isFormRequest(final MimeType contentType) {
+            return "application".equals(contentType.getPrimaryType()) &&
+                    "x-www-form-urlencoded".equals(contentType.getSubType());
+        }
+
+        private byte[] reconstructFormBody(final ServletRequest request) {
+            return request.getParameterMap().entrySet().stream()
+                    .flatMap(entry -> Arrays.stream(entry.getValue())
+                            .map(value -> encode(entry.getKey()) + "=" + encode(value)))
+                    .collect(joining("&"))
+                    .getBytes(UTF_8);
+        }
+
+        private static String encode(final String s) {
+            return RemoteRequest.encode(s, "UTF-8");
+        }
+
+    }
+
+    @AllArgsConstructor
+    private static abstract class Streaming implements State {
+
+        @Getter(PROTECTED)
+        private final ByteArrayInputStream stream;
+
+        @Override
+        public ServletInputStream getInputStream(final ServletRequest request) {
+            return new ServletInputStreamAdapter(stream);
+        }
+
+    }
+
+    private static final class Buffering extends Streaming {
+
+        private final byte[] body;
+
+        Buffering(final byte[] body) {
+            this(body, new ByteArrayInputStream(body));
+        }
+
+        Buffering(final byte[] body, final ByteArrayInputStream stream) {
+            super(stream);
+            this.body = body;
+        }
+
+        @Override
+        public State without() {
+            return new Ignoring(body);
+        }
+
+        @Override
+        public byte[] getBody() {
+            return body;
+        }
+
+    }
+
+    private static final class Ignoring extends Streaming {
+
+        private byte[] body;
+
+        Ignoring(final byte[] body) {
+            super(new ByteArrayInputStream(body));
+            this.body = body;
+        }
+
+        @Override
+        public State with() {
+            return new Buffering(body, getStream());
+        }
+
+    }
+
+    private static final class Passing implements State {
+
+    }
+
+    private final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Fields should be declared at the top of the class, before any method declarations, constructors, initializers or inner classes.](https://app.codacy.com/app/whiskeysierra/logbook/pullRequest?prid=4506178)